File: src/main/java/org/example/eventhub/entity/Ticket.java
package org.example.eventhub.entity;

import jakarta.persistence.*;
import java.math.BigDecimal;
import lombok.*;
import org.example.eventhub.enums.TicketStatus;

@Entity
@NoArgsConstructor
@Getter
@Setter
@AllArgsConstructor
@Builder
public class Ticket {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    private Long id;

    @ManyToOne(optional = false)
    @JoinColumn(name = "event_id")
    private Event event;

    @ManyToOne(optional = false)
    @JoinColumn(name = "user_id")
    private User user;

    @ManyToOne(optional = false)
    @JoinColumn(name = "order_id")
    private Order order;

    @Column(nullable = false)
    @Builder.Default
    private BigDecimal price = BigDecimal.ZERO;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private TicketStatus status = TicketStatus.RESERVED;
}

File: src/main/kotlin/org/example/eventhub/entity/Ticket.kt
package org.example.eventhub.entity

import jakarta.persistence.*
import org.example.eventhub.enums.TicketStatus
import java.math.BigDecimal

@Entity
@Table(name = "tickets")
class Ticket(

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    var id: Long? = null,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "event_id", nullable = false)
    var event: Event,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    var user: User,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "order_id", nullable = false)
    var order: Order,

    @Column(nullable = false)
    var price: BigDecimal,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: TicketStatus = TicketStatus.RESERVED
) {

    fun pay() {
        status = TicketStatus.PAID
    }

    fun cancel() {
        status = TicketStatus.CANCELLED
    }
}
